client/grpc/tmservice/service.go

```diff
@@ -129,27 +114,31 @@ func (s queryServer) GetValidatorSetByHeight(ctx context.Context, req *GetValida
 	if req.Height > chainHeight {
 		return nil, status.Error(codes.InvalidArgument, "requested block height is bigger then the chain length")
 	}
+	r := &GetValidatorSetByHeightResponse{}
+	r.BlockHeight, r.Validators, err = validatorsOutput(s.clientCtx, &req.Height, page, limit)
+	return r, err
+}
 
-	validatorsRes, err := rpc.GetValidators(s.clientCtx, &req.Height, &page, &limit)
-
+func validatorsOutput(ctx client.Context, height *int64, page, limit int) (int64, []*Validator, error) {
+	vsRes, err := rpc.GetValidators(ctx, height, &page, &limit)
 	if err != nil {
-		return nil, err
+		return 0, nil, err
 	}
 
-	outputValidatorsRes := &GetValidatorSetByHeightResponse{
-		BlockHeight: validatorsRes.BlockHeight,
-		Validators:  make([]*Validator, len(validatorsRes.Validators)),
-	}
-
-	for i, validator := range validatorsRes.Validators {
-		outputValidatorsRes.Validators[i] = &Validator{
-			Address:          validator.Address,
-			ProposerPriority: validator.ProposerPriority,
-			PubKey:           validator.PubKey,
-			VotingPower:      validator.VotingPower,
+	out := make([]*Validator, len(vsRes.Validators))
+	for i, v := range vsRes.Validators {
+		pkBz, err := codec.MarshalIfcJSON(ctx.JSONMarshaler, v.PubKey)
+		if err != nil {
+			return 0, nil, err
+		}
+		out[i] = &Validator{
+			Address:          v.Address,
+			ProposerPriority: v.ProposerPriority,
+			PubKey:           string(pkBz),
```

Comment:
I just noticed that string here is not ideal, I created https://github.com/cosmos/cosmos-sdk/issues/8054

Reply:
Commented in the task. Let's handle it there.